### ISSUE TITLE
problems with related resources example

### ISSUE THREAD

**Issue by ttdonovan** [michiel/jsonapi-rust#35]
I'm having issues getting an example of related resources working. You can find my example here:

**Branch:** https://github.com/ttdonovan/jsonapi-rust/tree/ttdonovan/example-related-resources

**Diff:** https://github.com/michiel/jsonapi-rust/compare/master...ttdonovan:ttdonovan/example-related-resources

According to https://jsonapi-validator.herokuapp.com/ the `data/related_resources.json` is a valid and compliant.

I'm wondering if the issue with the "relationships" and having the correct models defined in Rust. If that is the case then should anything not defined just be ignored like an "attributes"?

Any insight or guidance would greatly be appreciated.

**Comment by ttdonovan** [comment]
Using the branch that includes a fix for #36 I was able to make more progress. However I am running into an thread overflowed its stack somewhere in `JsonApiModel::resource_to_attrs`.

```
$ cargo run --example related_resources
   Compiling jsonapi v0.5.4 (file:///Users/tanner/Workspace/ttdonovan-github/jsonapi-rust)
    Finished dev [unoptimized + debuginfo] target(s) in 3.9 secs
     Running `target/debug/examples/related_resources`
Example of Related Resources: /v1/authors/1?include=books.chapters

thread 'main' has overflowed its stack
fatal runtime error: stack overflow
Abort trap: 6
```
I will continue to do some more investigating.

**Comment by michiel** [comment]
I'm running your example with the #36 changes merged against latest master and it works for me. Can you rebase the latest changes and try again?

    jsonapi-rust(ttdonovan/example-related-resources) $ git merge tddonovan/api-relationship-data-optional
    jsonapi-rust(ttdonovan/example-related-resources) $ RUST_BACKTRACE=1 cargo run --example related_resources ; echo $?
        Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
         Running `target/debug/examples/related_resources`
    Example of Related Resources: /v1/authors/1?include=books.chapters
    Ok(JsonApiDocument { data: Some(Single(Resource { _type: "authors", [... removed content ...]

    0

**Comment by ttdonovan** [comment]
@michiel I just pushed the revised example to my remote branch. You'll see the issue is with calling `let author = Author::from_jsonapi_document(&data).unwrap();`

**Comment by ttdonovan** [comment]
I think I've started to nail it down to something in the recursion of `Self::resource_to_attrs` for a `Some(IdentifierData::Single(ref identifier))`. My example will run if apply the following change:

```
$ git diff
diff --git a/examples/related_resources.rs b/examples/related_resources.rs
index b12c96b..3f7f3f6 100644
--- a/examples/related_resources.rs
+++ b/examples/related_resources.rs
@@ -11,7 +11,7 @@ struct Author {
     #[serde(skip_serializing_if = "Option::is_none")]
     books: Option<Vec<Book>>,
 }
-jsonapi_model!(Author; "author"; has many books);
+jsonapi_model!(Author; "author"); // ; has many books
 
 #[derive(Debug, Serialize, Deserialize)]
 struct Book {
@@ -36,5 +36,5 @@ fn main() {
     let json = include_str!("../data/related_resources.json");
     let data: JsonApiDocument = serde_json::from_str(&json).unwrap();
     let author = Author::from_jsonapi_document(&data).unwrap();
-    println!("foobar");
+    println!("{:?}", author);
 }
diff --git a/src/model.rs b/src/model.rs
index 267ebc9..decfeb2 100644
--- a/src/model.rs
+++ b/src/model.rs
@@ -153,12 +153,12 @@ pub trait JsonApiModel: Serialize
                 for (name, relation) in relations {
                     let value = match relation.data {
                         Some(IdentifierData::None) => Value::Null,
-                        Some(IdentifierData::Single(ref identifier)) => {
-                            let found = Self::lookup(identifier, inc)
-                                .map(|r| Self::resource_to_attrs(r, included) );
-                            to_value(found)
-                                .expect("Casting Single relation to value")
-                        },
+                        // Some(IdentifierData::Single(ref identifier)) => {
+                        //     let found = Self::lookup(identifier, inc)
+                        //         .map(|r| Self::resource_to_attrs(r, included) );
+                        //     to_value(found)
+                        //         .expect("Casting Single relation to value")
+                        // },
                         Some(IdentifierData::Multiple(ref identifiers)) => {
                             let found: Vec<Option<ResourceAttributes>> =
                                 identifiers.iter().map(|id|{
@@ -170,6 +170,7 @@ pub trait JsonApiModel: Serialize
                                 .expect("Casting Multiple relation to value")
                         },
                         None => Value::Null,
+                        _ => Value::Null,
                     };
                     new_attrs.insert(name.to_string(), value);
                 }
```

**Comment by ttdonovan** [comment]
Here is what I believe is happening - not sure how to correct this at this time but this example demonstrates the issue.

Book with ID 1 has a relationship with Chapters
call to `resource_to_attrs` for Chapter with ID 1 has a relationship with Book
call to `resource_to_attrs` for Book with ID 1 again with relationship with Chapters
then program panics

Here is `println!("resource_to_attrs: {:?}\n\n", resource);!` at start of `resource_to_attrs` before panic.

```
resource_to_attrs: Resource { _type: "books", id: "1", attributes: {"date_published": String("1954-07-29"), "updated_at": String("2018-02-14 00:31:18"), "title": String("The Fellowship of the Ring"), "created_at": String("2018-02-14 00:31:18")}, relationships: Some({"author": Relationship { data: Some(Single(ResourceIdentifier { _type: "authors", id: "1" })), links: Some({"self": String("/v1/authors/1/relationships/author"), "related": String("/v1/authors/1/author")}) }, "photos": Relationship { data: None, links: Some({"self": String("/v1/authors/1/relationships/photos"), "related": String("/v1/authors/1/photos")}) }, "chapters": Relationship { data: Some(Multiple([ResourceIdentifier { _type: "chapters", id: "1" }, ResourceIdentifier { _type: "chapters", id: "2" }, ResourceIdentifier { _type: "chapters", id: "3" }, ResourceIdentifier { _type: "chapters", id: "4" }, ResourceIdentifier { _type: "chapters", id: "5" }, ResourceIdentifier { _type: "chapters", id: "6" }, ResourceIdentifier { _type: "chapters", id: "7" }, ResourceIdentifier { _type: "chapters", id: "8" }, ResourceIdentifier { _type: "chapters", id: "9" }, ResourceIdentifier { _type: "chapters", id: "10" }, ResourceIdentifier { _type: "chapters", id: "11" }, ResourceIdentifier { _type: "chapters", id: "12" }, ResourceIdentifier { _type: "chapters", id: "13" }, ResourceIdentifier { _type: "chapters", id: "14" }, ResourceIdentifier { _type: "chapters", id: "15" }, ResourceIdentifier { _type: "chapters", id: "16" }, ResourceIdentifier { _type: "chapters", id: "17" }, ResourceIdentifier { _type: "chapters", id: "18" }, ResourceIdentifier { _type: "chapters", id: "19" }, ResourceIdentifier { _type: "chapters", id: "20" }, ResourceIdentifier { _type: "chapters", id: "21" }, ResourceIdentifier { _type: "chapters", id: "22" }])), links: Some({"self": String("/v1/authors/1/relationships/chapters"), "related": String("/v1/authors/1/chapters")}) }, "firstChapter": Relationship { data: None, links: Some({"self": String("/v1/authors/1/relationships/firstChapter"), "related": String("/v1/authors/1/firstChapter")}) }, "series": Relationship { data: Some(Single(ResourceIdentifier { _type: "series", id: "1" })), links: Some({"self": String("/v1/authors/1/relationships/series"), "related": String("/v1/authors/1/series")}) }, "stores": Relationship { data: None, links: Some({"related": String("/v1/authors/1/stores"), "self": String("/v1/authors/1/relationships/stores")}) }}), links: Some({"self": String("/v1/authors/1")}), meta: None }


resource_to_attrs: Resource { _type: "chapters", id: "1", attributes: {"created_at": String("2018-02-14 00:31:18"), "ordering": Number(1), "updated_at": String("2018-02-14 00:31:18"), "title": String("A Long-expected Party")}, relationships: Some({"book": Relationship { data: Some(Single(ResourceIdentifier { _type: "books", id: "1" })), links: Some({"self": String("/v1/authors/1/relationships/book"), "related": String("/v1/authors/1/book")}) }}), links: Some({"self": String("/v1/authors/1")}), meta: None }


resource_to_attrs: Resource { _type: "books", id: "1", attributes: {"date_published": String("1954-07-29"), "updated_at": String("2018-02-14 00:31:18"), "title": String("The Fellowship of the Ring"), "created_at": String("2018-02-14 00:31:18")}, relationships: Some({"author": Relationship { data: Some(Single(ResourceIdentifier { _type: "authors", id: "1" })), links: Some({"self": String("/v1/authors/1/relationships/author"), "related": String("/v1/authors/1/author")}) }, "photos": Relationship { data: None, links: Some({"self": String("/v1/authors/1/relationships/photos"), "related": String("/v1/authors/1/photos")}) }, "chapters": Relationship { data: Some(Multiple([ResourceIdentifier { _type: "chapters", id: "1" }, ResourceIdentifier { _type: "chapters", id: "2" }, ResourceIdentifier { _type: "chapters", id: "3" }, ResourceIdentifier { _type: "chapters", id: "4" }, ResourceIdentifier { _type: "chapters", id: "5" }, ResourceIdentifier { _type: "chapters", id: "6" }, ResourceIdentifier { _type: "chapters", id: "7" }, ResourceIdentifier { _type: "chapters", id: "8" }, ResourceIdentifier { _type: "chapters", id: "9" }, ResourceIdentifier { _type: "chapters", id: "10" }, ResourceIdentifier { _type: "chapters", id: "11" }, ResourceIdentifier { _type: "chapters", id: "12" }, ResourceIdentifier { _type: "chapters", id: "13" }, ResourceIdentifier { _type: "chapters", id: "14" }, ResourceIdentifier { _type: "chapters", id: "15" }, ResourceIdentifier { _type: "chapters", id: "16" }, ResourceIdentifier { _type: "chapters", id: "17" }, ResourceIdentifier { _type: "chapters", id: "18" }, ResourceIdentifier { _type: "chapters", id: "19" }, ResourceIdentifier { _type: "chapters", id: "20" }, ResourceIdentifier { _type: "chapters", id: "21" }, ResourceIdentifier { _type: "chapters", id: "22" }])), links: Some({"self": String("/v1/authors/1/relationships/chapters"), "related": String("/v1/authors/1/chapters")}) }, "firstChapter": Relationship { data: None, links: Some({"self": String("/v1/authors/1/relationships/firstChapter"), "related": String("/v1/authors/1/firstChapter")}) }, "series": Relationship { data: Some(Single(ResourceIdentifier { _type: "series", id: "1" })), links: Some({"self": String("/v1/authors/1/relationships/series"), "related": String("/v1/authors/1/series")}) }, "stores": Relationship { data: None, links: Some({"related": String("/v1/authors/1/stores"), "self": String("/v1/authors/1/relationships/stores")}) }}), links: Some({"self": String("/v1/authors/1")}), meta: None }

# output stops because of panic
```

**Comment by ttdonovan** [comment]
I've created a pull request that exposes this bug with `cargo test` #37

**Comment by JadedEvan** [comment]
Closed in favor of changes presented in #72
